tokenizers/src/models/unigram/trainer.rs

```diff
@@ -257,55 +264,87 @@ impl UnigramTrainer {
 
+        let mut vsum = 0.0;
+        let mut freq: Vec<f64> = vec![0.0; pieces.len()];
+        let mut inverted: Vec<Vec<usize>> = vec![Vec::new(); pieces.len()];
+
         let bos_id = pieces.len() + 1;
         let eos_id = pieces.len() + 2;
 
-        // First, segments the current sentencepieces to know
-        // how each sentencepiece is resegmented if this sentencepiece is removed
-        // from the vocabulary.
-        // To do so, we take the second best segmentation of sentencepiece[i].
-        // alternatives[i] stores the sequence of second best sentencepieces.
-        for (id, (token, _score)) in pieces.iter().enumerate() {
-            // Always keep unk.
-            if id == 0 {
-                always_keep[id] = false;
-                continue;
-            }
-            let mut lattice = Lattice::from(token, bos_id, eos_id);
-            model.populate_nodes(&mut lattice);
-
-            let nbests = lattice.nbest(2);
-            if nbests.len() == 1 {
-                always_keep[id] = true;
-            } else if nbests[0].len() >= 2 {
-                always_keep[id] = false;
-            } else if nbests[0].len() == 1 {
-                always_keep[id] = true;
-                for node in &nbests[1] {
-                    let alt_id = node.borrow().id;
-                    alternatives[id].push(alt_id);
-                }
-            }
-        }
+        let (tx, rx) = mpsc::channel();
 
-        // Second, segments all sentences to compute likelihood
-        // with a unigram language model. inverted[i] stores
-        // the set of sentence index where the sentencepieces[i] appears.
-        let mut vsum = 0.0;
-        let mut freq: Vec<f64> = vec![0.0; pieces.len()];
-        let mut inverted: Vec<Vec<usize>> = vec![Vec::new(); pieces.len()];
-        // TODO reparallelize this
-        for (i, (sentence, count)) in sentences.iter().enumerate() {
-            let mut lattice = Lattice::from(sentence, bos_id, eos_id);
-            model.populate_nodes(&mut lattice);
-            vsum += *count as f64;
-            for node_ref in lattice.viterbi() {
-                let id = node_ref.borrow().id;
-                freq[id] += *count as f64;
-                inverted[id].push(i);
-            }
+        let tx2 = tx.clone();
+
+        enum Message {
+            AlwaysKeep((usize, bool)),
+            Alternatives((usize, usize)),
+            Vsum(f64),
+            Freq((usize, f64)),
+            Inverted((usize, usize)),
         }
 
+        rayon::scope(|s| {
+            // First, segments the current sentencepieces to know
+            // how each sentencepiece is resegmented if this sentencepiece is removed
+            // from the vocabulary.
+            // To do so, we take the second best segmentation of sentencepiece[i].
+            // alternatives[i] stores the sequence of second best sentencepieces.
+            s.spawn(|_| {
```

Comment:
using [rayon::spawn](https://docs.rs/rayon/1.5.1/rayon/fn.spawn.html) which makes it possible for `tx & rx` work asynchronously